### PR TITLE
IGNITE-17521 Fixes after review.

### PR DESCRIPTION
**Pull request by sergeyuttsel**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/lang/Function4.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.ignite.lang;
+
+import java.util.Objects;
+import java.util.function.Function;
+
+/**
+ * Represents a function that accepts three arguments and produces a result.
```
Comment (sanpwc) [inline modules/core/src/main/java/org/apache/ignite/lang/Function4.java:24]: typo, three -> four

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/exception/PrimaryReplicaMissException.java [hunk 2]
```diff
@@ -30,29 +29,27 @@ public class PrimaryReplicaMissException extends IgniteInternalException {
...
      * The constructor.
      *
-     * @param requestAddress Primary replica address from request.
-     * @param localAddress Local primary replica address.
      * @param requestTerm Raft term from request.
      * @param localTerm Local raft term.
      */
-    public PrimaryReplicaMissException(NetworkAddress requestAddress, NetworkAddress localAddress, Long requestTerm, Long localTerm) {
-        this(requestAddress, localAddress, requestTerm, localTerm, null);
+    public PrimaryReplicaMissException(Long requestTerm, Long localTerm) {
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/exception/PrimaryReplicaMissException.java:35]: Nope, it's still required to know expectedPrimaryReplicaAdress and currentPrimaryReplicaAddress. Expected is the one that was used for routing - in other words recipient localAddress and currentPrimaryReplicaAddress is leader's address that was returned by leaderAndTerm().
By the way, what do u think about renaming params to
expectedPrimaryReplicaAdress
currentPrimaryReplicaAddress
expectedPrimaryReplicaTerm
currentPrimaryReplicaTerm
together with using one-param-on-a-line formatting?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 3]
```diff
@@ -136,21 +138,43 @@ public PartitionReplicaListener(
...
-            processScanCloseAction((ReadWriteScanCloseReplicaRequest) request);
-
-            return CompletableFuture.completedFuture(null);
+        CompletableFuture<Object> fut;
+        if (request instanceof ReadWriteReplicaRequest) {
+            ReadWriteReplicaRequest request0 = (ReadWriteReplicaRequest) request;
+
+            fut = raftClient.refreshAndGetLeaderWithTerm()
+                    .thenCompose((replicaAndTerm) -> {
+                        if (!replicaAndTerm.get2().equals(request0.term())) {
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:147]: It's also required to check whether localNode is the same as the current leader.

File: modules/core/src/main/java/org/apache/ignite/lang/Function4.java (hunk 1 above)
Comment (sergeyuttsel) [reply modules/core/src/main/java/org/apache/ignite/lang/Function4.java]: fixed

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replication/request/ReadWriteReplicaRequest.java [hunk 4]
```diff
@@ -28,4 +28,13 @@
...
     UUID transactionId();
+
+    /**
+     * Gets a raft term.
+     * TODO: A temp solution until lease-based engine will be implemented (IGNITE-17256, IGNITE-15083)
+     *
+     * @return Gets a raft term.
+     */
+    @Marshallable
+    Long term();
```
[the remark this answers is not in the document]
Comment (sergeyuttsel) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replication/request/ReadWriteReplicaRequest.java:40]: fixed

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/exception/PrimaryReplicaMissException.java (hunk 2 above)
Comment (sergeyuttsel) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/exception/PrimaryReplicaMissException.java]: Fixed, but expectedPrimaryReplicaAdress and currentPrimaryReplicaAddress are not needed.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java (hunk 3 above)
Comment (sergeyuttsel) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java]: Fixed, but we need to check term only.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 5]
```diff
@@ -992,4 +998,36 @@ private CompletableFuture<RowId> takeLocsForReplace(BinaryRow searchKey, BinaryR
...
+        return raftClient.refreshAndGetLeaderWithTerm()
+                .thenCompose(replicaAndTerm -> {
+                            Long currentTerm = replicaAndTerm.get2();
+
+                            Long expectedTerm = null;
+
+                            if (request instanceof ReadWriteReplicaRequest) {
+                                expectedTerm = ((ReadWriteReplicaRequest) request).term();
+                            } else if (request instanceof TxFinishReplicaRequest) {
+                                expectedTerm = ((TxFinishReplicaRequest) request).term();
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:1018]: TxCleanupReplicaRequest is missing, so please also add term to it.

**Pull request closed without merging**